### PR TITLE
ModCompat changes

### PR DESCRIPTION
**Pull request by Cheaterpaul**

### CODE REVIEW

**Comment by Cheaterpaul** [comment]
make sure the changes are binary compatible

**Comment by maxanier** [comment]
Make sure to tag anything that could be solved in a better way with breaking changes with a `TODO 1.16`

File: src/api/java/de/teamlapen/vampirism/api/entity/factions/IFactionRegistry.java [hunk 1]
```diff
@@ -77,6 +77,19 @@
...
+     *
+     * @param id              Faction id e.g. for level command
+     * @param entityInterface Interface all entities implement
+     * @param color           Color e.g. for level rendering
+     * @param villageFactionData       village capture related utility class (if null will gets filled with dummy)
+     * @param <T>             Interface all entities implement
+     * @return The created faction
+     */
+    @ThreadSafeAPI
+    <T extends IFactionEntity> IFaction registerFaction(ResourceLocation id, Class<T> entityInterface, Color color, boolean hostileTowardsNeutral, @Nullable IVillageFactionData villageFactionData);
```
Comment (Cheaterpaul) [inline src/api/java/de/teamlapen/vampirism/api/entity/factions/IFactionRegistry.java:91]: could also be handled like the chat color, with a setter method.

File: src/api/java/de/teamlapen/vampirism/api/entity/CaptureEntityEntry.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package de.teamlapen.vampirism.api.entity;
+
+import net.minecraft.entity.EntityType;
+import net.minecraft.entity.MobEntity;
+import net.minecraft.util.WeightedRandom;
+
+public class CaptureEntityEntry extends WeightedRandom.Item {
```
Comment (Cheaterpaul) [inline src/api/java/de/teamlapen/vampirism/api/entity/CaptureEntityEntry.java:7]: could be moved into IVillageFactionData

File: src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java [hunk 3]
```diff
@@ -39,19 +74,22 @@
          * Random existing villager in totemTile
          * Used as a "seed" villager to get a valid spawn point.
          */
-        private final @Nonnull
-        VillagerEntity seedVillager;
+        private final @Nullable
+        MobEntity oldEntity;
```
Comment (Cheaterpaul) [inline src/api/java/de/teamlapen/vampirism/api/event/VampirismVillageEvent.java:78]: acceptable because the event was never fired

**Comment by Cheaterpaul** [comment]
Not sure about the events. As you said, if they were never fired, Binary compatibility is not necessary.

File: src/api/java/de/teamlapen/vampirism/api/entity/factions/IVillageFactionData.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        @Override
+        public VillagerProfession getFactionVillageProfession() {
+            return VillagerProfession.NONE;
+        }
+
+        @Override
+        public List<CaptureEntityEntry> getCaptureEntries() {
+            return ImmutableList.of();
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/entity/factions/IVillageFactionData.java]: `Collections.emptyList()`?

File: src/api/java/de/teamlapen/vampirism/api/entity/player/actions/DefaultAction.java [hunk 5]
```diff
@@ -10,16 +10,8 @@
  * Default implementation for an action
  */
 public abstract class DefaultAction<T extends IFactionPlayer> extends ForgeRegistryEntry<IAction> implements IAction {
-    private final IPlayableFaction<T> faction;
     private String translationKey;
 
-    /**
-     * @param faction
-     */
-    public DefaultAction(IPlayableFaction<T> faction) {
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/entity/player/actions/DefaultAction.java:19]: This is still breaking.
Keep constructor and `getFaction` method.
Add additional empty constructor, that initializes with null and override `getFaction` in subclass. Add TODO

File: src/api/java/de/teamlapen/vampirism/api/entity/player/skills/DefaultSkill.java [hunk 6]
```diff
@@ -18,21 +16,10 @@
...
 
     private final Map<IAttribute, AttributeModifier> attributeModifierMap = new HashMap<>();
-    private final IPlayableFaction<T> faction;
     private int renderRow;
     private int renderColumn;
     private String translationKey;
 
-    protected DefaultSkill(IPlayableFaction<T> faction) {
-        this.faction = faction;
-    }
```
Comment (maxanier) [inline src/api/java/de/teamlapen/vampirism/api/entity/player/skills/DefaultSkill.java]: Same as with the action

**Review by maxanier — CHANGES_REQUESTED** [review]

**Pull request merged**